Ghost Constraint Stabilization

juliand665 opened the pull request:
As part of [my thesis](https://ethz.ch/content/dam/ethz/special-interest/infk/chair-program-method/pm/documents/Education/Theses/Julian_Dunskus_MS_Description.pdf) under @Aurel300, I've been writing specs for the standard library (#1249), which will ship with Prusti for users to benefit from.

## Ghost Constraints → Type-Conditional Spec Refinements

These specs often rely on ghost constraints, e.g. to provide a way to opt out of specs [applied by default](https://github.com/viperproject/prusti-dev/pull/1249/files#diff-b5e647a76ab0c8608a5d7a0ab8ff619d6bbf84217f854e134d5ff60b636c0e0aR5-R7) to certain trait implementations, or to [specify known return values](https://github.com/viperproject/prusti-dev/pull/1249/files#diff-b4d918ba4e39891439964fe0debf2b28f0d9cef6e4417bb0e9321cd34d5b6fb1R9) for e.g. `core::mem::size_of`. Since these specs will be active by default, they would require enabling ghost constraints to use Prusti at all (barring the option of conditionalizing specs on enabled Prusti features somehow). Hence we propose to stabilize ghost constraints, addressing some issues with them to get them to a place where this makes sense.

Ghost constraints are now called type-conditional spec refinements (`TypeCondSpec` for short) and spelled as follows:
```rs
#[refine_spec(where Self: Copy, T: Copy, [
    pure,
    ensures(result === T::default())
])
```
The `where` draws a clear analogy to existing Rust syntax to help users understand this attribute, and it even results in correct syntax highlighting for the constraints. Note also that `pure` is now supported here (correctly—my previous implementation was broken), and that you can apply multiple bounds.

## Extern Spec Improvements

I also polished up & resolved some limitations with extern specs:

- An error is now thrown when extern specs are applied to methods with bodies, rather than stubs.
- They can now be applied to trait impls with generic arguments, which was previously cautiously rejected. There is [a unit test](https://github.com/viperproject/prusti-dev/pull/1271/files#diff-0eda14ab21130110aee12255a5b0425d332f42439cd49d31acd0ab6b4c2408a5) ensuring that this works correctly.
- Extern specs on traits now support bounds in a `where` clause.
- They can now be applied to free functions, taking an argument to specify that function's module path, e.g. `#[extern_spec(core::mem)] fn swap(...)`. This argument also works on traits & modules (`impl`s already allow qualifying involved type paths).

## Miscellaneous

As part of this PR, I enabled auto-formatting for the `prusti-specs` crate, and reformatted it in the process. This crate is one of the last few holdovers that weren't yet auto-formatted, the other big one being `prusti-viper`. It looked like this crate is not currently being majorly worked on by anyone else!

Finally, I fixed a bug where lifetimes would make trait method resolution fail, by instead erasing all regions involved in the resolution process. This made some tests fail before I [reworked](https://github.com/viperproject/prusti-dev/commit/b3be7e81ed2cfb16eaa007f7431b5a65b1b38318) the piece of code that encodes lifetimes for function calls. This bug made it impossible to specify e.g. `impl Add<&i32> for i32`, because Prusti could not resolve the general `Add::add` to this specific implementation (due to the lifetimes necessarily differing between call site & declaration site).

juliand665 committed:
recognize and reject extern specs which define bodies for their methods

juliand665 committed:
support multiple constraints in ghost constraints

juliand665 committed:
correctly support purity ghost constraints

turns out the previous implementation was flawed in that it made the function pure whether or not the constraint held

juliand665 committed:
require at least one bound for ghost constraints

juliand665 committed:
report example structure only when it makes sense

juliand665 committed:
add compiletests for ghost constraint purity

juliand665 committed:
auto-format prusti-specs crate

juliand665 committed:
unify handling of bodies in extern specs and check more places

juliand665 committed:
add unit test for bodies in extern specs

juliand665 committed:
update unit tests

juliand665 committed:
parse new syntax for ghost constraints/conditional spec refinements

juliand665 committed:
add failing test for extraneous comma before specs

juliand665 committed:
update example in error note

juliand665 committed:
remove ghost constraint feature flag

you can still introduce unsound behavior with them, but they're far from the only way to do so, and absolutely necessary for prusti std

juliand665 committed:
rename ghost constraints to conditional spec refinements in user-facing strings & comments

juliand665 committed:
formatting

juliand665 committed:
address clippy's concerns

juliand665 committed:
add unit test for trait impls with generic arguments

removed a check preventing use of this feature previously; this effectively formalizes it as working

juliand665 committed:
improve test

juliand665 committed:
desugaring fixes

expressions are not implicitly in parentheses, creating very strange errors when desugaring e.g. a + b to &a + b

juliand665 committed:
conditional spec refinements -> type-conditional spec refinements

juliand665 committed:
rename ghost constraints to type-conditional spec refinements everywhere

juliand665 committed:
update test to reflect that generics are supported

not thoroughly tested, but let's not let perfect be the enemy of good enough

juliand665 committed:
update test to reflect that generics are supported

not thoroughly tested, but let's not let perfect be the enemy of good enough

juliand665 committed:
adjust join_spans testing-specific behavior

testing-specific workaround is now only present when cfg(test) is active

juliand665 committed:
make typecondspec parser more robust, require comma again

juliand665 committed:
update overlooked holdover of old ghost constraint term

juliand665 committed:
fix typo

juliand665 committed:
update compiletests to new typecondspec syntax

juliand665 committed:
Merge remote-tracking branch 'upstream/master'

juliand665 committed:
update preparser tests to include the new parentheses

juliand665 committed:
update cargo test

juliand665 committed:
update more compiletests

juliand665 committed:
update more compiletests

juliand665 committed:
allow merge clauses for extern trait specs

implemented by simply tacking on our bound to the existing clause

juliand665 committed:
allow extern_spec to take a module argument & apply to free functions

the module argument is not valid on impls, since those already work with qualified type paths

juliand665 committed:
update tests

juliand665 committed:
allow trailing commas in where clauses on extern specs on traits

juliand665 committed:
ignore lifetimes in pure function substs

juliand665 committed:
ignore lifetimes in extern spec sanity check too

juliand665 committed:
always use lifetimes from args instead of generics

generic lifetimes are erased during trait impl resolution

juliand665 committed:
clippy

juliand665 committed:
update user guide with new typecondspec syntax

juliand665 committed:
add regression test for trait resolution with references

juliand665 committed:
update broken link

Aurel300 commented on docs/user-guide/src/verify/type_cond_spec.md:
```diff
@@ -0,0 +1,37 @@
+
+Thus, any client implementing `Eq` on a custom type can take advantage of the additional semantics of the total equivalence.
+
+> Refinements like these are not scoped. Therefore, considering the previous example, implementing `Eq` on a type implies that the total equivalence contract is always considered on the type, irrespective of whether `Eq` is in scope or not.
```
I find this a bit confusing/redundant: implementation of traits is not scoped in general. At the callsite, the constraint can be resolved either way for sure, because the `impl` exists either on the concrete type or on the trait (+ blanket `impl`s).

juliand665 replied:
Yeah, I was a bit torn on this comment, so I just tried to keep things unchanged from the previous state. But you're absolutely right: the compiler knows about the impl with or without an import; that would only be relevant when actually trying to use one of its members. Do you reckon we can just get rid of the quoted part entirely?

Aurel300 replied:
Yes, I would drop this.

juliand665 committed:
use lifetimes from substs when possible

juliand665 committed:
erase lifetimes only when resolved method differs

Aurel300 commented on prusti-contracts/prusti-specs/src/extern_spec_rewriter/impls.rs:
```diff
@@ -340,16 +329,28 @@ mod tests {
+                    #[prusti::extern_spec = "trait_impl"]
+                    #[allow(unused, dead_code)]
+                    #[prusti::trusted]
+                    fn foo(_self: &mut MyStruct, arg1: I) {
+                        <MyStruct as MyTrait<I>> :: foo :: <>(_self, arg1)
```
Where is the type parameter `I` declared after rewriting? Is it somehow part of `newtype_ident`?

juliand665 replied:
I did a double take here too originally—`I` is ([was](https://github.com/juliand665/prusti-dev/commit/76e6f6598f6161d0afa5ce23ce0fa03ebfe90a73)) in fact not a type parameter! Hence why it's not where you'd expect. I think it makes sense to test both a blanket `impl<I>` and a specific `impl` though. I'll update the tests accordingly, and rename `I` in the single case to not break conventions ;)

Aurel300 commented on prusti-contracts/prusti-specs/src/extern_spec_rewriter/traits.rs:
```diff
@@ -49,29 +59,42 @@ fn generate_new_struct(item_trait: &syn::ItemTrait) -> syn::Result<GeneratedStru
-    let self_where_clause: syn::WhereClause = parse_quote! {
-        where #self_type_ident: #self_type_trait
-    };
+    let self_where_clause: syn::WhereClause =
```
Maybe worth checking: is a `trait Foo where Self: Bar ...` clause handled correctly, i.e. is the `Self` rewritten to `Prusti_T_Self` as well?

juliand665 replied:
Good catch! It indeed does not. Funnily enough it appears it would still work, if not for the fact that `Self` does not appear to be getting rewritten in pre-/postconditions anyway, so this area might need some more love in general. I'll handle the where clauses though.

juliand665 replied:
On second thought, I'm not sure what I needed where clauses in extern specs for anymore? I don't have any code lying around using it, and it doesn't seem useful, since you don't need them to figure out which trait it is, and it seems you already get the type bounds for "free", i.e. you're able to use methods from `Bar` in the specs for `Foo` like this:

```rs
#[extern_spec]
trait Foo {
    #[ensures(result == self.bar())]
    fn foo(&self) -> i32;
}
```

(I initially wrote this as static methods rather than taking `&self`, bringing the `Self`-in-specs issue to my attention.)

Aurel300 replied:
Ah right, you're right: the bounds on `Self` should of course be inherited from the actual trait definition. Never mind!

Aurel300 commented on prusti-contracts/prusti-specs/src/extern_spec_rewriter/traits.rs:
```diff
@@ -203,15 +233,16 @@ impl ProvidedTypeParam {
         match path.segments[0].ident.to_string().as_str() {
             "generic" => Some(ProvidedTypeParam::GenericType(clone_without_attrs())),
```
We should get rid of this mechanism altogether, we only want to allow "generic" generics.

juliand665 replied:
Ooh yeah, we can entirely get rid of the whole annotation mess thanks to typecondspecs turning stable! Always satisfying to delete code.

Aurel300 commented on prusti-contracts/prusti-specs/src/rewriter.rs:
```diff
@@ -164,6 +170,30 @@ impl AstRewriter {
+            #[allow(unused_must_use, unused_parens, unused_variables, dead_code)]
+            #[prusti::spec_only]
+            #[prusti::spec_id = #spec_id_str]
+            fn #item_name() {} // we only care about this for evaluating
```
"for evaluating" what? The `where` clauses I guess?

juliand665 replied:
Oh yeah, I must have gotten distracted halfway through that comment. I'll replace it with something clearer.

Aurel300 commented on prusti-contracts/prusti-specs/src/specifications/mod.rs:
```diff
@@ -11,7 +11,7 @@
 /// ensures ::= "ensures", "(", prusti_expr, ")" ;
-/// 
+///
 /// This grammar doesn't yet handle the unintuitive precedence difference between `&&` and `||` operators.
```
(This entire comment is outdated now.)

juliand665 replied:
Not just because of this PR though, right? Should I just put down a FIXME or something?

Aurel300 replied:
No, I just noticed because of the reformat. But feel free to just remove the comment, it's not valuable if it's not accurate anyway.

Aurel300 commented on prusti-contracts/prusti-specs/src/specifications/preparser.rs:
```diff
@@ -905,14 +946,16 @@ impl RustOp {
-        // This works even when compiled with stable
-        s1.unwrap().join(s2.unwrap()).expect("Failed to join spans!").into()
-    } else {
+    // Tests don't run in the proc macro context, so this gets a little funky for them
```
@JonasAlaif I think I recall we talked about this at some point. Does this `cfg` solution look good to you? Did you try this before the `catch_unwind` one?

JonasAlaif replied:
I guess it's late now anyway, but this solution looks better yeah

Aurel300 commented on prusti-interface/src/specs/typed.rs:
```diff
@@ -203,6 +203,7 @@ pub struct ProcedureSpecification {
     pub pledges: SpecificationItem<Vec<Pledge>>,
     pub trusted: SpecificationItem<bool>,
     pub terminates: SpecificationItem<Option<LocalDefId>>,
+    pub purity: SpecificationItem<Option<DefId>>, // for type-conditional spec refinements
```
I found this addition really strange. Why do we need another field just for purity? Isn't the refinability of `kind` already expressed in the `SpecificationItem` type?

Aurel300 commented on prusti-interface/src/specs/typed.rs:
```diff
@@ -436,6 +438,23 @@ impl SpecGraph<ProcedureSpecification> {
+                let constrained_spec = self.get_constrained_spec_mut(constraint);
+                constrained_spec.kind =
+                    SpecificationItem::Inherent(ProcedureSpecificationKind::Pure);
+                // need to store this as well, since without pres or posts we couldn't find any def id with the trait bounds
```
Is this the answer to the comment about the `purity` field? Is it because we still won't generate a spec item *just* for purity?

juliand665 replied:
It is indeed! From what I understand, the [current design](https://github.com/viperproject/prusti-dev/blob/a4ac54bbb747607ca9ab5d24f0013202a2e6e171/prusti-viper/src/encoder/mir/specifications/constraints.rs#L218-L221) (which we deemed out of scope to overhaul) looks at all the spec items, gathers `where` clauses as `ParamEnv`s, checks that there's only one such env (since we currently only support a single typecondspec refinement for any method), and returns it.

The problem is that to do so, it [runs through](https://github.com/viperproject/prusti-dev/blob/a4ac54bbb747607ca9ab5d24f0013202a2e6e171/prusti-viper/src/encoder/mir/specifications/constraints.rs#L234) all the pre- and postconditions listed in the `ProcedureSpecification`. I didn't touch this code initially until I realized it fails for ghost constraints that only establish purity and nothing else (no pres or posts) because the loop has nothing to run through. So [my solution](https://github.com/juliand665/prusti-dev/blob/0684867cc8765c8cc1abee796882393dd5761603/prusti-viper/src/encoder/mir/specifications/constraints.rs#L235) is to store and iterate over the purity spec item as well.

Aurel300 commented on prusti-tests/tests/parse/ui/and.stdout:
```diff
@@ -23,7 +23,7 @@ use prusti_contracts::*;
 #[prusti::spec_id = "$(NUM_UUID)"]
 fn prusti_pre_item_test1_$(NUM_UUID)() -> bool {
-    !!((true && true): bool)
+    !!(((true) && (true)): bool)
```
Where did we get the extra parentheses from?

juliand665 replied:
They came from [here](https://github.com/viperproject/prusti-dev/pull/1271/commits/d4069dd6e86805754cfa48090f88e08bac2090e8), fixing issues if e.g. `lhs` were `foo || bar`. They are technically avoidable, but I think that would require quite some restructuring to apply initial parentheses to any "untrusted" tokens, then respect precedence levels in the further desugaring. Probably not worth it?

Aurel300 replied:
I see the issue now. Yes, we could somehow detect when the parentheses are not needed, but that's indeed more effort than it's worth.

Aurel300 commented on prusti-tests/tests/verify/fail/ghost-constraints/purity.rs:
```diff
@@ -0,0 +1,12 @@
+#[trusted]
+fn test<T>(_t: T) -> bool { true }
+
+#[derive(Clone, PartialEq, Eq)]
```
Unnecessary derive?

Aurel300 commented on prusti-tests/tests/verify/pass/extern-spec/module-arg.rs:
```diff
@@ -0,0 +1,77 @@
+
+#[extern_spec(module)]
+mod inner {
+    use crate::*;
```
I wonder: do we really need to leave these imports in? Do we keep the `mod` structure around when rewriting this kind of extern spec? Or is this just to appease IDEs?

juliand665 replied:
`#[extern_spec]` [erases](https://github.com/viperproject/prusti-dev/blob/a4ac54bbb747607ca9ab5d24f0013202a2e6e171/prusti-contracts/prusti-contracts-proc-macros/src/lib.rs#L73-L77) whatever it's attached to when not building for prusti, so I don't think this helps IDEs. I believe this is actually necessary because we're currently still preserving the module structure rather than moving everything to the top level. Thus, the `prusti_contracts` attributes are not in scope, and if you do import them, this case still gives an error because it can't resolve `module`, which is also not in scope within `inner`.

Aurel300 commented on prusti-tests/tests/verify/pass/type-cond-specs/purity.rs:
```diff
@@ -0,0 +1,9 @@
+fn test<T>(_t: T) -> bool { true }
+
+fn main() {
+	assert!(test(5) == test(5));
```
Indented with a tab.

Aurel300 commented on prusti-tests/tests/verify/ui/failing-postcondition.stderr:
```diff
@@ -1,8 +1,8 @@
- --> $DIR/failing-postcondition.rs:8:11
+ --> $DIR/failing-postcondition.rs:8:28
   |
 8 | #[ensures(something_true() && false)]
```
What made the span change here?

juliand665 replied:
Not sure how that one slipped through! I couldn't find it by just looking so it looks like I'll have to throw in some print statements to track it down.

juliand665 replied:
Oh great, looks like [this change](https://github.com/viperproject/prusti-dev/pull/1271/commits/d4069dd6e86805754cfa48090f88e08bac2090e8) messed that up somehow. That doesn't make sense to me because it's already using an explicit span. Any ideas?

Aurel300 commented on prusti-tests/tests/verify_overflow/fail/type-cond-specs/in-non-trusted-function.rs:
```diff
@@ -0,0 +1,11 @@
+trait A {}
+
+#[refine_spec(where T: A, [
+	ensures(true)
```
tab

juliand665 replied:
Non-autoformatted files will be my undoing…

Aurel300 commented on prusti-viper/src/encoder/mir/procedures/encoder/mod.rs:
```diff
@@ -1488,11 +1488,9 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
         // The called method might be a trait method.
         // We try to resolve it to the concrete implementation
         // and type substitutions.
+        let query = self.encoder.env().query;
```
Why this alias? It doesn't seem to be used later on.

juliand665 replied:
Without it, the long line spilled over into a silly-looking chained function call spanning like 5 lines, lol. I considered updating the later usage, but that felt worse. Should I just accept the other formatting rather than polluting this long function's namespace?

juliand665 committed:
tiny fix

got lost in the refactor; i blame copilot lol

juliand665 committed:
fix some test indentation

juliand665 committed:
remove unnecessary derive

juliand665 committed:
rename missed test folders

juliand665 committed:
improve generic trait tests

juliand665 committed:
update comment

juliand665 committed:
drop unhelpful comment

juliand665 committed:
remove outdated comment

better no information than inaccurate information!

juliand665 committed:
fix spans for binary ops

while preserving the safety of parenthesizing incoming LHS/RHS

juliand665 committed:
obsolete #[generic]/#[concrete] for traits

the latter didn't work anyway, and now that we have type-cond-specs, we know we won't need it, so we can drop the annotations entirely.

juliand665 committed:
update tests for trait parameter attribute removal

juliand665 committed:
update preparser cfg tests

juliand665 committed:
rewrite self type in newly-allowed where clauses

where clauses on traits

juliand665 committed:
bump cache version

we were getting spurious failures for the core proof tests from old data

juliand665 committed:
mangle function names for extern specs

juliand665 committed:
flatten extern spec modules

it's like they were never there! no more use crate::*;

juliand665 committed:
update tests

juliand665 committed:
only mangle free functions

keep the changes to a minimum

juliand665 committed:
Merge branch 'master' into master
